### PR TITLE
JSON schema for flink-job

### PR DESCRIPTION
**Pull request by iblutrifork**
Add JSON schema for flink-job

Continuation of https://github.com/trifork/cheetah-charts/pull/95
Implementing https://jira.trifork.com/browse/KAMDP-341

### CODE REVIEW

File: charts/flink-job/values.schema.json [hunk 1]
```diff
@@ -0,0 +1,590 @@
...
+        "jarURI": {
+          "type": "string"
+        },
+        "entryClass": {
+          "type": "string"
+        },
+        "args": {
+          "type": "array",
+          "items": {}
+        },
```
Comment (AndersBennedsgaard) [inline charts/flink-job/values.schema.json:216]:
```suggestion
        "args": {
          "type": "array",
          "items": {
            "type": "string"
          }
        },
```

File: charts/flink-job/values.schema.json [hunk 2]
```diff
@@ -0,0 +1,590 @@
...
+        "topics": {
+          "type": "array",
+          "items": {}
+        },
+        "parallelism": {
+          "type": "integer"
+        },
+        "state": {
+          "type": "string"
```
Comment (AndersBennedsgaard) [inline charts/flink-job/values.schema.json:225]: Must be either "running" or "suspended"

File: charts/flink-job/values.schema.json [hunk 3]
```diff
@@ -0,0 +1,590 @@
...
+              "type": "string"
+            },
+            "selectors": {
+              "type": "object"
+            },
+            "labels": {
+              "type": "object"
+            }
+          },
+          "required": ["enabled", "labels", "selectors", "targetPort"]
```
Comment (AndersBennedsgaard) [inline charts/flink-job/values.schema.json:372]:
```suggestion
          "required": ["enabled", "targetPort"]
```

File: charts/flink-job/values.schema.json [hunk 4]
```diff
@@ -0,0 +1,590 @@
...
+        "automountServiceAccountToken": {
+          "type": "boolean"
+        }
+      },
+      "required": [
+        "annotations",
+        "automountServiceAccountToken",
+        "create",
+        "name"
+      ]
```
Comment (AndersBennedsgaard) [inline charts/flink-job/values.schema.json:398]:
```suggestion
      "required": [
        "automountServiceAccountToken",
        "create"
      ]
```

File: charts/flink-job/values.schema.json [hunk 5]
```diff
@@ -0,0 +1,590 @@
...
+        },
+        "resource": {
+          "type": "object",
+          "properties": {
+            "memory": {
+              "type": "string"
+            },
+            "cpu": {
+              "type": "number"
+            }
```
Comment (AndersBennedsgaard) [inline charts/flink-job/values.schema.json:537]:
```suggestion
            "memory": {
              "oneOf": [
                {
                  "type": "string"
                },
                {
                  "type": "number"
                }
              ]
            },
            "cpu": {
              "oneOf": [
                {
                  "type": "string"
                },
                {
                  "type": "number"
                }
              ]
            }
```

File: charts/flink-job/values.schema.json [hunk 6]
```diff
@@ -0,0 +1,590 @@
...
+          },
+          "required": ["cpu", "memory"]
```
Comment (AndersBennedsgaard) [inline charts/flink-job/values.schema.json:539]:
```suggestion
          }
```

File: charts/flink-job/values.schema.json [hunk 7]
```diff
@@ -0,0 +1,590 @@
...
+          "items": {}
+        },
+        "volumes": {
+          "type": "array",
+          "items": {}
+        },
+        "volumeMounts": {
+          "type": "array",
+          "items": {}
+        },
```
Comment (AndersBennedsgaard) [inline charts/flink-job/values.schema.json:560]: Same as above.

File: charts/flink-job/values.schema.json [hunk 8]
```diff
@@ -0,0 +1,590 @@
...
+        "initContainers",
+        "podAnnotations",
+        "podLabels",
+        "podTemplate",
+        "ports",
+        "replicas",
+        "resource",
+        "volumeMounts",
+        "volumes"
+      ]
```
Comment (AndersBennedsgaard) [inline charts/flink-job/values.schema.json:587]:
```suggestion
      "required": [
        "env",
        "envFrom",
        "podAnnotations",
        "podTemplate",
        "ports",
        "replicas",
        "resource",
        "volumeMounts",
        "volumes"
      ]
```

**Review by AndersBennedsgaard — CHANGES_REQUESTED** [review]
It would be nice to be able to remove many/all of the `fail` and `required` fields in https://github.com/trifork/cheetah-charts/blob/main/charts/flink-job/templates/_helpers.tpl.
Such as https://github.com/trifork/cheetah-charts/blob/998e0953da1a893685a35c7a53f4e69a8582121b/charts/flink-job/templates/_helpers.tpl#L145 which says that `topics[*].name` is required, and https://github.com/trifork/cheetah-charts/blob/998e0953da1a893685a35c7a53f4e69a8582121b/charts/flink-job/templates/_helpers.tpl#L151 which says that `topics[*].type` must be `input` or `output`

File: charts/flink-job/values.schema.json [hunk 9]
```diff
@@ -0,0 +1,590 @@
...
+      "properties": {
+        "log4j-console.properties": {
+          "type": "string"
+        }
+      },
+      "required": ["log4j-console.properties"]
+    },
+    "mode": {
+      "type": "string"
+    },
```
[the remark this answers is not in the document]
Comment (iblutrifork) [reply charts/flink-job/values.schema.json:87]: Yup, I'll use an enum here

**Review by AndersBennedsgaard — COMMENTED** [review]
Small change, but otherwise LGTM :smile:

**Review by AndersBennedsgaard — APPROVED** [review]

**Pull request merged**